### ISSUE TITLE
Recent Objects - storing all metadata exceeds cookie character limit

### ISSUE THREAD

**Issue by cmlewis** [tsgrp/hpi#690]
When recent objects is configured (esp with 10 objects displaying), the object name, id, and all metadata is stored in the resulting cookie. With long object names and a few metadata columns configured, we quickly reach the max number of characters for a cookie (Chrome specifically).

We should only store object ids, then do a fetch for each and get the configured metadata. We should also remove the object from the cookie if it doesn't exist anymore.

**Comment by tbrandstatter** [comment]
Changed the recent object dashlet to use OCOs for properties instead of saving them on the cookie. The only thing on the cookie now are the object id and the container id (for wizard).

Fixed with HPI revision #12458.

**Comment by cmlewis** [comment]
Thanks @tbrandstatter! When we fetch each OCO, are we removing the object id from the cookie if it no longer exists?

**Comment by tbrandstatter** [comment]
@cmlewis yes, we added that as well!

**Comment by cmlewis** [comment]
Great, thank you!

Thanks,
Christy
(312) 899-2947

**Comment by gsteimer** [comment]
@tbrandstatter - can you check the sorting of the folders in the cookie?  In the below screenshot, the 'Misc Subfolder' was the last folder I opened, but it's ordered in the middle of the list.  Recent Objects should be ordered with the most recent object at the top and the oldest one falling off the list.

![image](https://cloud.githubusercontent.com/assets/1377228/8209261/b26ba31c-14d1-11e5-8dbb-a35dbbb5070d.png)

**Comment by tbrandstatter** [comment]
Modified recentObjects so it sorts the folders properly--since now have ocos, we had to create a loop to make sure the order of items on the cookie would match the order of the objects being displayed onto the view.

Fixed with HPI revision #12474

**Comment by cmlewis** [comment]
All, I started up HPI with a different Alfresco (meaning my object ids stored in the cookie don't exist in the current Alfresco I started up). Shouldn't this new logic prevent any errors from displaying by removing those non-existent object ids from the cookie? If the logic is checking if the node is 'deleted,' it really should be checking if it exists or not, therefore preventing these errors:
```
com.tsgrp.opencontent.core.exception.OCRuntimeException: Failed to get properties: com.tsgrp.opencontent.core.exception.OCRuntimeException:
The node ref workspace://SpacesStore/f88b6bea-ef3b-47a5-bcea-1a239e5e490e does not exist.
       at com.tsgrp.opencontent.alfresco.content.AlfrescoEmbContentImpl.enhancedGetProperties(AlfrescoEmbContentImpl.java:515)
       at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
       at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
       at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
       at java.lang.reflect.Method.invoke(Unknown Source)
       at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:317)
       at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
```

![image](https://cloud.githubusercontent.com/assets/4852302/8255856/9cd701de-1668-11e5-9717-e0b09cc03ebc.png)

**Comment by cmlewis** [comment]
Quick update - this is odd but I seem to get this same OC exception even after I clear out my cookies.

**Comment by tbrandstatter** [comment]
unable to replicate error on alfresco or documentum

**Comment by aviking** [comment]
@cmlewis The error that you logged from the tomcat is an expected error in this situation because as you said the objectId's dont exist in the second repository. I believe the modal that you saw (based on the screen shot) is comming from a different dashlet (savedsearchdashlet line 576). If you can confirm this then I think we can close this issue.

**Comment by cmlewis** [comment]
@aviking @tbrandstatter I can't confirm currently where the modal was coming from (on the road), so it's definitely possible it's coming from a different dashlet.

But I think there might be a misunderstanding about the deletion. Part of this enhancement is to remove any items from the cookie that no longer exist in the repository. At a high level, the logic should be this:
1. Fetch OCO.
2a. OCO found, display properties.
2b. OCO not found, remove from cookie.

So, by changing repositories, we GUARANTEE the OCOs will not be found. Therefore they should be removed. It's fairly easy to test - start up HPI on repo 1 and click around to update the cookie with recent objects. Shutdown, startup with repo 2. Load dashboard - all items should be removed because they can't be found in repo 2. If this scenario can be tested successfully, we can close this issue.

**Comment by gsteimer** [comment]
@cmlewis - I'm closing this since we've recently implemented #384 - recent objects are now stored in the repository in user prefs, so the issue of recent objects trying to fetch objects in a different repository is now moot.

Feel free to comment and re-open if you disagree.
